Better UX for updates

philandstuff opened the pull request:
Currently, when you update a service with this broker, it assumes
you're passing in the whole configuration, which can be surprising:

    # create a queue with a dead letter queue policy
    cf create-service myservice -c '{redrive_max_receive_count: 5}'
    # update retention period, oops I removed the DLQ policy
    cf update-service -c '{message_retention_period: 1000000}'

This PR fixes that behaviour by making it CloudFormation's problem.  That is, we turn everything that a user might update into a CloudFormation parameter, and ask CloudFormation to preserve the existing value if the user didn't specify anything.

risicle commented:
Soo... if we're doing this here, should we look at doing the same for _bind_ parameters?

philandstuff commented:
This is ready for review.  I removed support for the `fifo` plan because it didn't seem to work, and hence also removed ContentBasedDeduplication because that's only valid for FIFO queues.

> Soo... if we're doing this here, should we look at doing the same for bind parameters?

Yes probably, presumably the same update issue affects bind parameters, and we should probably be consistent in how we define our templates.  But let's get this PR merged as-is first, then I can do the user template after?

philandstuff commented:
Worth noting something I haven't written down yet:

This changes update operations such that updates *only* change parameters on existing stacks, they *never* change the underlying template.  This feels like a nice property: things that are updatable should be parameters, not in the template.

philandstuff commented:
argh triple post, but worth expanding on what I wrote ☝️ 

> Yes probably, presumably the same update issue affects bind parameters, and we should probably be consistent in how we define our templates. But let's get this PR merged as-is first, then I can do the user template after?

In other words, I think we should merge this, merge #11, then open another PR to make #11 look like this one (possibly with @risicle and me pairing).

chrisfarms commented on sqs/provider.go:
```diff
@@ -90,6 +103,13 @@ func (s *Provider) Provision(ctx context.Context, provisionData provideriface.Pr
+func mkParameter(name string, value int) *cloudformation.Parameter {
+	return &cloudformation.Parameter{
+		ParameterKey:   aws.String(name),
+		ParameterValue: aws.String(fmt.Sprint(value)),
```
style: could we use `strconv.Itoa` or a `"%d"` formatter here if this is intended to always be an `int` (go vet and tools can enforce types for the format strings, but can't help here)... or change the mkParamater argument to be an interface if it is intended to be a generic stringify-anything-ifier? .... looks like it's in a middleground at the moment.

risicle commented:
Sounds good to me.

Just to make a note as to what we were talking about out of band, note that binding parameters aren't updateable. The only use I could see for storing the parameters for a binding is to be able to correctly handle the distinction between `200 OK` responses and `409 Conflict` responses to bind requests.

And yeah, maybe we're not going to actually get into that stuff anyway.

chrisfarms reviewed (commented):
This looks good, I changes to Provider are an improvement and I like the "templates are pretty static"-ness again.

I'm about as happy as I expected with a big YAMLy blob :smile: ... I have concerns about not being able to lean on any kind of typing when making changes to the template and things like whitespace now being important ... but nothing I haven't raised already.
  
I think the positional placeholder would be my biggest complaint, they really do make the template hard to read and require you to know your format strings, it might not immediately be obvious (unlike a helm/smarty/go style `{{ thing }}`)

philandstuff commented on sqs/provider.go:
```diff
@@ -47,38 +49,49 @@ type Provider struct {
+	if provisionData.Plan.Name == "fifo" {
+		return nil, apiresponses.NewFailureResponseBuilder(errors.New("FIFO plan unimplemented"), http.StatusNotImplemented, "not-implemented").WithEmptyResponse().Build()
+	}
+	tmplParams := TemplateParams{}
```
I actually like the separation between “updatable” and “non updatable” parameters. Then we can have confidence by design that Update() operation won’t accidentally destroy/recreate something. As a bonus, the Booleans (which are painful to make cloudformation template parameters) are non-updatable.

chrisfarms replied:
yeah I can see the benefit :+1:  ... maybe there is something simple we could do to make it easier for "future me" to intuit this distinction, like naming them `immutableParams` vs `userParams` or something (think my concern with the split is the next person coming along to make a change having to make the decision "is this a template-params or is this a parameter-param") .... this is obviously a pretty minor point

philandstuff replied:
Yeah the existing names are bad and I like your proposed ones

philandstuff replied:
I had a brainwave and realised if we turn the "non updatable" params into a QueueTemplateBuilder struct then it all becomes much clearer.

risicle commented:
This is a mad idea but do you think it's worth keeping a "version" parameter among the CF params to make life easier if/when we need to make changes and grandfather old service instances?

philandstuff commented:
@risicle oof, that's a good point, will think about it.

risicle commented:
Alternatively, add it when we make the first change, and take its absence as "version 0"

philandstuff commented:
Yeah the underlying problem is "how do we manage multiple versions of templates created by multiple versions of sqs-broker" and storing a version in the template somehow is one solution (another is to tag the stack rather than putting something in the template, another is to somehow migrate all existing templates when we change the meta-template).  It's probably a discussion we should have more broadly but I'm coming to think it doesn't live in this PR since it was a preexisting problem.

risicle commented:
:+1: I forgot about tags
